### ISSUE TITLE
Tracking issue for RFC 2363, "Allow arbitrary enums to have explicit discriminants"

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#60553]
This is a tracking issue for the RFC "Allow arbitrary enums to have explicit discriminants" (rust-lang/rfcs#2363).

**Steps:**

- [x] Implement the RFC – https://github.com/rust-lang/rust/pull/60732
- [ ] Adjust documentation ([see instructions on rustc-guide][doc-guide])
- [x] Stabilization PR – #86860

[stabilization-guide]: https://rust-lang.github.io/rustc-guide/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rust-lang.github.io/rustc-guide/stabilization_guide.html#documentation-prs

**Unresolved questions:**

1. [ ] Should discriminants of enumerations with fields be specified as variant attributes?
2. [ ] Should this apply only to enumerations with an explicit representation?

**Comment by eddyb** [comment]
Mentoring-wise: I'd suggest trying out an example and then tracking down where the error comes from. I don't remember, it's either during parsing or in `ast_validation`.

Most of the compiler should already support this, where doing the uniform thing is easier.
The "niche" optimization might assume variants with any data have consecutive discriminants, but it should be easy to fix, and we can figure it out after an initial PR is open.

**Comment by jswrenn** [comment]
Has anyone stepped up to implement this? If not, I'd like to take a stab.

The current error arises in parsing. I haven't fiddled with it yet, but I suspect a good first step is to remove the `else` before this `if`: https://github.com/rust-lang/rust/blob/33cde4aac2ec1b0a493d0acaa4c4fb45de0c6e94/src/libsyntax/parse/parser.rs#L7781-L7789

**Comment by oli-obk** [comment]
> Has anyone stepped up to implement this?

I don't know of anyone

>  If not, I'd like to take a stab.

Go for it!

> I haven't fiddled with it yet, but I suspect a good first step is to remove the else before this if: 

That looks like the right start, yes.

**Comment by jswrenn** [comment]
I've started work on this [here](https://github.com/jswrenn/rust/tree/arbitrary_enum_discriminant) by [removing the parsing restriction preventing non-unit variants from having explicit discriminants](https://github.com/jswrenn/rust/commit/ff93aed9022145beb1a37dc618d0db4735e0df18). As you predicted @eddyb, things seem to basically just work; e.g.:
```rust
#![feature(core_intrinsics)]

enum Enum {
  A(u16) = 7,
  B(u32) = 42,
}

fn main() {
  use std::intrinsics::discriminant_value;

  unsafe {
    assert_eq!(discriminant_value(&Enum::A(1)), 7);
    assert_eq!(discriminant_value(&Enum::B(2)), 42);
  }
}
```

As it stands, adding this feature required removing more lines of code than it involved writing new code! :D

Future work that comes to mind include writing tests and putting this behind a feature flag. Shall I go ahead and file an initial PR?

**Comment by oli-obk** [comment]
Sweet! Yes please open a PR with the feature gate and throw some tests in for good measure (there may already be tests checking for the "discriminator values can only be used with a field-less enum" error message, those should now be emitting a feature gate error instead)

**Comment by Centril** [comment]
The `#![feature(arbitrary_enum_discriminant)]` was implemented on 2019-06-26 in https://github.com/rust-lang/rust/pull/60732 which was written by @jswrenn and reviewed by @pnkfelix and others.

**Comment by jswrenn** [comment]
## Documentation
> * [ ]  Adjust documentation ([see instructions on rustc-guide](https://rust-lang.github.io/rustc-guide/stabilization_guide.html#documentation-prs))

I've submitted a PR to update the reference to reflect this feature (in its current state): rust-lang-nursery/reference#639.

I do not believe that either the Book or _Rust By Example_ need updating:
1. The Book [does not mention discriminants at all](https://doc.rust-lang.org/nightly/book/ch06-01-defining-an-enum.html).
2. _Rust By Example_ [mentions C-like enums](https://doc.rust-lang.org/rust-by-example/custom_types/enum/c_like.html), but does not go into their memory layout. Since this feature is only relevant to advance users who care about the memory layout of enums, _Rust By Example_ is probably ill-suited for the topic.

## Unresolved Questions

> 1. [ ]  Should discriminants of enumerations with fields be specified as variant attributes?

Presently: no. Discriminants of all types of enumerations are specified with `=` followed by a constant expression. I implemented this approach because doing so was simplest, but I'd also argue this is the _right_ approach since it means the syntax for specifying discriminants is consistent across variant types.

> 2. [ ]  Should this apply only to enumerations with an explicit representation?

Presently: yes, this feature applies only to enumerations using a primitive representation. Since this feature is only relevant when you care about the precise memory layout of enums, this feature is only reasonable to use when the precise layout of an enum is well-specified.

**Comment by newpavlov** [comment]
What are layout guarantees for such enums? Ideally I would like to be able to write code likes this:
```rust
#[repr(u8, C)]
enum Foo {
    A = 0,
    B { a: u16 } = 1,
    C { a: u32, b: u16 } = 2,
}
```
And I want for this type to have exactly the same memory layout as:
```rust
#[repr(C)]
struct foo_t {
    pub discr: u8,
    pub payload: foo_u,
}

union foo_u {
    b: u16,
    c: c_t,
}

#[repr(C)]
struct c_t {
    a: u32,
    b: u16,
}
```
Meaning if `repr(C)` is used:
- Discriminant should always be stored in the beginning.
- Discriminant can not be used for niche-optimizations.
- Variant fields should follow `repr(C)`.

Having such guarantees would allow in some cases some neat zero-cost translations of C types to safe Rust ones.

**Comment by eddyb** [comment]
@newpavlov You get most of that (the placement and size of the discriminant) from just `#[repr(u8)]`, I'm not sure if `#[repr(C)]` adds more to that.

Keep in mind that since your discriminants are the default ones (starting at `0` and adding `1` for each variant), you don't need this feature!

We defined all these back in https://github.com/rust-lang/rfcs/pull/2195, for Servo <-> Firefox I believe.

**Comment by elichai** [comment]
What about casting those enums into the primitive? 
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=2e69ca93ace1d4983bd9782074e2df6a

And than even `#[repr(Rust)]` support

**Comment by sztomi** [comment]
Being a rust newbie, I instinctively tried this and the compiler sent me to this issue. I see this is fairly old, but I'm not familiar with how quickly features make it into a stable version of the compiler. Is it possible to tell how far this feature is from stabilizing?

**Comment by eddyb** [comment]
> I see this is fairly old

It's not even an year old, and within that, not much has happened once this was implemented.
There's https://github.com/rust-lang/rust/issues/60553#issuecomment-511235271 - which looks like stabilization could've been attempted, but nobody tried?

**Comment by nikomatsakis** [comment]
The next step to stabilizing would be to prepare a **stabilization report**, which basically summarizes: 

* history of the feature -- when was it implemented, how long has it been on nightly, is anybody using it?
* what the feature does (very briefly, we don't have to duplicate the RFC)
* any changes since the RFC
* examples of test cases showing the behavior
* answers to any unresolved questions from the RFC
* any other interesting things that came up

**Comment by Centril** [comment]
For an example of a stabilization report, see e.g., https://github.com/rust-lang/rust/pull/67712 and please reach out to e.g., myself (Discord or Zulip works) before publishing it.

**Comment by jswrenn** [comment]
There's still work to be done to document this feature before stabilization—[my PR](https://github.com/rust-lang/reference/pull/639) updating the reference is stalled.

The main roadblock I hit was I wasn't aware of [RFC2195](https://github.com/dkl/rust-rfcs/blob/master/text/2195-really-tagged-unions.md) when I submitted that PR, and the details of that RFC aren't already well-documented in the reference. Since it's trivial to explain what this feature does in the context of RFC2195, a PR should probably first be submitted to the reference updating the [Type Layout](https://doc.rust-lang.org/reference/type-layout.html) chapter to document RFC2195.

[There's some hairy terminology issues nobody's been particularly consistent about, too](https://github.com/rust-lang/reference/pull/639#issuecomment-511515290).

**Comment by Centril** [comment]
@jswrenn Note that we do not require that you update the reference before stabilization. It's perfectly acceptable to write a good report and then have that report be the material for later changes to the book, the reference, the rustc-dev-guide, etc. I would be happy to work with you towards a stabilization report! (and ❤️ for the work you've done so far).

**Comment by elpiel** [comment]
Is there any progress on this RFC?

**Comment by jswrenn** [comment]
@elpiel I think I recall seeing some chatter about this interacting with const generics (or resolving the names of constants?) but I'd have to go searching for it.

**Comment by Robbepop** [comment]
Has this potential optimization allowed by this feature has already been discussed somewhere?
https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=e40e4866f6a7864d462ebe9f4d07fa29

tldr; If I did not oversee anything important the compiler is able to fuse discriminants of nested `enum`s if it knows that the discriminants of all variants do NOT overlap. See the playground for an example.

**Comment by eddyb** [comment]
So I was wondering why `#[repr(inttype)]` is required and came across https://github.com/rust-lang/rust/pull/60732#issuecomment-491880240.

I believe this fails to account for e.g. `mem::Discriminant`, and also usecases where you *don't* need to extract the discriminant as a tag at all, just to specify the value.

**Comment by apps4uco** [comment]
Hi, I was just wondering if my use case would be supported by the fix

https://gist.github.com/e0ef47efa0bb5b3865cf5ddcc3eae92d

The idea is that supported enums are represented by known enum tags and 
other values are represented by Unknown(u16) which has the discriminant of the u16 parameter.

This would be useful for FFI to and from other languages,
and in my case from extensible file formats.

~~~
use num_enum::FromPrimitive;
#[derive(Debug, Eq, PartialEq, FromPrimitive)]
#[repr(u16)]
pub enum TagEnum {
    Unknown(u16),
    ImageWidth = 0x0100,
...etc
~~~

**Comment by fee1-dead** [comment]
## Stabilization Report

### Summary

Enables a user to specify *explicit* discriminants on arbitrary enums.

Previously, this was hard to achieve:

```rust
#[repr(u8)]
enum Foo {
    A(u8) = 0,
    B(i8) = 1,
    C(bool) = 42,
}
```

Someone would need to add 41 hidden variants in between as a workaround with implicit discriminants.

In conjunction with [RFC 2195](https://github.com/rust-lang/rfcs/blob/master/text/2195-really-tagged-unions.md), this feature would provide more flexibility for FFI and unsafe code involving enums.

### Test cases

Most tests are in [`src/test/ui/enum-discriminant`](https://github.com/rust-lang/rust/tree/master/src/test/ui/enum-discriminant), there are two [historical](https://github.com/rust-lang/rust/blob/master/src/test/ui/parser/tag-variant-disr-non-nullary.rs) [tests](https://github.com/rust-lang/rust/blob/master/src/test/ui/parser/issue-17383.rs) that are now covered by the feature.

### Edge cases

The feature is well defined and does not have many edge cases. 
One [edge case](https://github.com/rust-lang/rust/issues/70509) was related to another unstable feature named `repr128` and is resolved.

### Previous PRs

The [implementation PR](https://github.com/rust-lang/rust/pull/60732) added documentation to the Unstable Book, [rust-lang/reference#639](https://github.com/rust-lang/reference/pull/639) is waiting on stabilization.

### Resolution of unresolved questions

The questions are resolved in https://github.com/rust-lang/rust/issues/60553#issuecomment-511235271.

**Comment by AsafFisher** [comment]
What about the following scenario?:
```rust
let cmd_id = 0;
match cmd_id {
            Command::ExecuteCMD => Command::ExecuteCMD{remote_path: a, args: b}
        };
```

**Comment by fee1-dead** [comment]
I don't think that code would work with this RFC, I think it is pretty clear that it just allows one to specify explicit discriminants of variants on arbitrary enums. 

For your case I think you might just define a separate, c-like enum so the match would look better, or you can just hard code those numbers, or you can use a macro.

**Comment by Mark-Simulacrum** [comment]
Reopening per https://github.com/rust-lang/rust/pull/89884.

**Comment by CeleritasCelery** [comment]
What will it take to get this back on track to be stabilized? Is there anything a normal person can do to help this along, or does it still need to be discussed by the T-lang team?

**Comment by nikomatsakis** [comment]
Cross-posting from #89234:

There have been a lot of PRs and questions concerning corner cases related to this RFC, for example:

* when one is allowed to use `as`
* when one is allowed to use a custom repr
* when one is allowed to NOT use a custom repr

We in @rust-lang/lang meeting would love it if we could review a single write-up that covers a complete design, approve it, and then land PRs that do the implementation work, rather than being cc'd piecemeal. Could somebody check whether rust-lang/reference#1055 is the write-up we are looking for and, if not, extend it so that it is? That would be much, much appreciated. 💜 

(cc @nbdd0121 @fee1-dead #89234 #88203)

**Comment by ehuss** [comment]
Sorry for the confusion and derailing this for a corner case. I can try to summarize with `arbitrary_enum_discriminant` as of 2021-11-30.

Currently, on stable, the following is allowed:

```rust
// An enum with only unit variants.
// `repr` is optional
// Allows `as` casting
// This is the only form that allows explicit discriminants.
enum UnitOnly {
    Foo = 2,
    Bar = 4,
    Baz = 8,
}

// This is allowed on stable.
let x = UnitOnly::Foo as u8;

// An enum with fieldless struct and tuple variants.
// `repr` is optional
// Allows `as` casting
// Does not allow explicit discriminants.
enum Fieldless {
    Tuple(),
    Struct{},
    Unit,
}

// This is allowed on stable.
assert_eq!(Fieldless::Tuple() as u8, 0);
assert_eq!(Fieldless::Struct{} as u8, 1);
assert_eq!(Fieldless::Unit as u8, 2);

// Enums with fields do not allow `as` casting or explicit discriminants.
```

The `arbitrary_enum_discriminant` feature extends to allow the following cases:

```rust
#![feature(arbitrary_enum_discriminant)]

// An enum with fieldless struct and tuple variants.
// `repr` is required to specify explicit discriminants.
// Allows `as` casting
#[repr(u8)]
enum FieldlessExplicit {
    Tuple() = 1,
    Struct{} = 3,
    Unit = 9,
}

// arbitrary_enum_discriminant allows `as` casting these
assert_eq!(FieldlessExplicit::Tuple() as u8, 1);
assert_eq!(FieldlessExplicit::Struct{} as u8, 3);
assert_eq!(FieldlessExplicit::Unit as u8, 9);

// An enum with fields.
// `repr` is required to specify explicit discriminants.
// Does not allow direct `as` casting.
#[repr(u8)]
enum WithFields {
    Tuple(i32) = 1,
    Struct{f: i32} = 3,
    Unit = 5,
}

// let x = WithFields::Tuple(1) as u8;  // ERROR: casting not allowed
// It requires unsafe pointer casting to get the discriminant.
unsafe {
    assert_eq!(*(&WithFields::Tuple(123) as *const WithFields as *const u8), 1);
    assert_eq!(*(&WithFields::Struct { f: 4 } as *const WithFields as *const u8), 3);
    assert_eq!(*(&WithFields::Unit as *const WithFields as *const u8), 5);
}
```

The concern was the difference between the `Fieldless` case and the `FieldlessExplicit` one. 

If you add explicit discriminants to a `Fieldless` enum, why do you now need to specify `repr`?  This also results in three styles of enums that need to be described in the documentation. The hope was that it could be reduced to two.

#88203 proposes that `FieldlessExplicit` does not require `repr` to stay consistent with current stable behavior. This is under the idea that adding explicit discriminants to a fieldless enum should not suddenly require also adding `repr`. This results in only two styles of enums to document.

#89234 proposes going in the other direction, where `FieldlessExplicit`, even with `repr`, does not allow `as` casting. There are some that feel that allowing `as` casting with `Fieldless` was a mistake, and https://github.com/rust-lang/rust/pull/89234#issuecomment-977263797 suggests that a warning should be added for that case. However, this still leaves three cases to document. 

I am not strictly opposed to having three styles to document. I just want this to be a decision that is clearly made.

**Comment by jswrenn** [comment]
> If you add explicit discriminants to a `Fieldless` enum, why do you now need to specify `repr`?

One of my goals in implementing this RFC was to help incrementally reform the design and use of enums to be more explicit and less implicit. As far as I know, it's always been the case that Rust has _implicitly_ assigned discriminants to the variants of enums. And, for probably just as long, it's been the case that Rust has only *let* you provide explicit discriminants in the case of enums with *only* unit-like variants. But soon, with this feature, you can be explicit even if the enum contains non-unit like variants.

However, there's another spot of implicit behavior in the design of enums: the type of the discriminants is *implicitly* `isize`, unless you specify otherwise. This is a foot-gun. It is *especially* a foot-gun on the kinds of enums that `feature(arbitrary_enum_discriminant)` affects. Consider that there are generally two possible reasons you'd want to manipulate the discriminants of variants:
1. you want `as`-casting to behave in a certain way
2. you've added an explicit `repr` to the enum, and you want to control the values of the tag bytes

For the kinds of enums that `feature(arbitrary_enum_discriminant)` affects, you are really *only* ever manipulating the discriminants for this second purpose: to control the values of tag bytes. Why am I so confident of this? *Because the discriminant values **cannot** be observed with `as`-casting!* You can't `as` cast this, e.g.:
```rust
#[repr(u8)]
enum WithFields {
    Tuple(i32) = 1,
    Struct{f: i32} = 3,
    Unit = 5,
}
```
Thus, all we've accomplished with explicit discriminants is controlling the values of tag bytes. There's basically currently no reason to set explicit discriminants on such enums unless you care about memory layout, in which case you also *really* ought to be clear on what the repr of the enum is. **Given this, it seemed prudent to require that an explicit `repr` is provided.**

There's just one exception to this rule:
> For the kinds of enums that `feature(arbitrary_enum_discriminant)` affects, you are really *only* ever manipulating the discriminants for this second purpose: to control the values of tag bytes. Why am I so confident of this? *Because the discriminant values **cannot** be observed with `as`-casting!*

It's enums like this:
```rust
/* #[repr(u8)] */
enum FieldlessExplicit {
    Tuple() /* = 1 */,
    Struct{} /* = 3 */,
    Unit /* = 9 */,
}
```
To the surprise of a lot of people, you can *already* as-cast fieldless enums (even if they're not C-like). You might, in theory, *want* to be able to affect the `as`-casting behavior of this enum by setting explicit discriminants but *not care about memory layout*. And you're totally out of luck, because we're going to make you write `repr(...)`. 

This is *such* a niche exception (most people didn't even know you could as-cast such enums to begin with!) that I did not feel that this exception posed any sort of meaningful objection to always requiring `repr(...)` on enums affected by this feature. 

---

So, given this:
> You might, in theory, *want* to be able to affect the `as`-casting behavior of this enum by setting explicit discriminants but *not care about memory layout*. And you're totally out of luck, because we're going to make you write `repr(...)`.

...what if we, like, just *didn't* make you write `repr(...)` in this case? Well, that's the behavior proposed by #88203. But is writing an explicit `repr(...)` in such cases ever *bad*? Are we actually harming users by making them write an explicit repr if they want to have explicit discriminants on fieldless-but-non-c-like enums? Is there any compelling reason why we *shouldn't* nudge you towards being more explicit about the repr in this case? **I don't think so!**

The *only* "bad" consequence of writing an explicit repr on an enum is that it currently *also* turns off niche-based optimizations. But niche-finding optimizations don't really apply to fieldless enums. (AFAIK)

As far as I can tell, basically the only case negatively impacted by requiring explicit reprs on fieldless-but-non-c-like enums are *fieldless-but-non-c-like-enums-with-only-one-variant*; e.g.:
```rust
/* #[repr(u8)] */
enum ReallyWeird {
  Tuple() /* = 2 */,
}
```
If you make the discriminants explicit in *this* case, it will have the unwelcome impact of making the enum 1 byte, rather than 0 bytes in size.

So, broadly: I think we should be encouraging explicit reprs *and* explicit discriminants. #88203 is counterproductive to this goal, and can't easily be reversed. Except in one *incredibly* niche case, virtually nothing is lost by having this requirement.

We have two options that support this goal:
1. #89234, which "solves" the issue (of setting discriminants simultaneously doing two different things on fieldless-but-non-c-like enums) by prohibiting `as`-casting on fieldless-but-non-c-like enums
2. do nothing

I'd be happy with either option, though prefer doing nothing. The only downsides of #89234 I foresee are that it slightly complicates the documentation around `as`-casting, and poses a *tiny* stability hazard that would discourage authors of fieldless-but-non-c-like enums from adding explicit discriminants. If we merge #89234, we can go back to doing nothing in the future.

Putting on my pragmatist hat: It's silly for something as unusual as fieldless-but-non-c-like enums to block stabilization. While I think doing nothing is the right choice, we should just merge #89234. If we regret it, we can *probably* revert it without breaking compat. And then, if we decide we *actually* wished we went with #88203, we can simply lift the `repr(...)` restriction on fieldless-but-non-c-like enums altogether. Merging #89234 gives us the most future flexibility.

**Comment by MatrixDev** [comment]
@jswrenn, I've never liked `as`-syntax for enums, but there should be at least some way to get those discriminants in the future if we can set those. are there any ideas?
It would be nice if `Discriminant<T>` returned by the `std::mem::discriminant()` had a numerical value or something.

**Comment by fee1-dead** [comment]
Enums with `repr(inttype)` have a stable repr, so it is safe to convert `*const SomeEnum` to `*const inttype` and dereference to get the discriminant value.

**Comment by MatrixDev** [comment]
@fee1-dead, dereferencing will involve `unsafe` code, won't it? It doesn't seems logical to require `unsafe` code just to get a discriminant.

**Comment by nbdd0121** [comment]
This is the state before #89234 is merged:

|                       | Unit-only (C-like) | Fieldless                     | With fields                   |
|-----------------------|--------------------|-------------------------------|-------------------------------|
| Implicit Discr w/o repr       | Stable, can cast   | Stable, can cast              | Stable                        |
| Implicit Discr w/ repr | Stable, can cast   | Stable, can cast              | Stable                        |
| Explicit Discr  w/o repr | Stable, can cast   | Not allowed | Not allowed |
| Explicit Discr w/ repr | Stable, can cast   | `arbitrary_enum_discriminant`, can cast |  `arbitrary_enum_discriminant` |

The current state is
|                       | Unit-only (C-like) | Fieldless                     | With fields                   |
|-----------------------|--------------------|-------------------------------|-------------------------------|
| Implicit Discr w/o repr       | Stable, can cast   | Stable, can cast              | Stable                        |
| Implicit Discr w/ repr | Stable, can cast   | Stable, can cast              | Stable                        |
| Explicit Discr w/o repr       | Stable, can cast   | Not allowed | Not allowed |
| Explicit Discr w/ repr | Stable, can cast   | `arbitrary_enum_discriminant` |  `arbitrary_enum_discriminant` |

Option: Reverting #89234 and applying #88203 gives:
|                       | Unit-only (C-like) | Fieldless                     | With fields                   |
|-----------------------|--------------------|-------------------------------|-------------------------------|
| Implicit Discr | Stable, can cast   | Stable, can cast              | Stable                        |
| Explicit Discr | Stable, can cast   | `arbitrary_enum_discriminant`, can cast |  `arbitrary_enum_discriminant` |

i.e. treat "fieldless enums" similar to unit-only enums.

Option: Deprecate fieldless casting gives
|                       | Unit-only (C-like) | Fieldless                     | With fields                   |
|-----------------------|--------------------|-------------------------------|-------------------------------|
| Implicit Discr w/o repr       | Stable, can cast   | Stable (casting warns)              | Stable                        |
| Implicit Discr w/ repr | Stable, can cast   | Stable (casting warns)              | Stable                        |
| Explicit Discr w/o repr       | Stable, can cast   | Not allowed | Not allowed |
| Explicit Discr w/ repr | Stable, can cast   | `arbitrary_enum_discriminant` |  `arbitrary_enum_discriminant` |

i.e. treat "fieldless enums" similar to enums with fields.

From what I perceive from Zulip discussions it seems that people generally think that `Variant()` is closer to `Variant(())` than `Variant` so "fieldless" enums shouldn't receive special treatment regarding castability, so we should go with the latter option.

Regardless which option to take, the current state (with #89234 merged) is the most restrictive combination that allows both options in the future, and should unblock `arbitrary_enum_discriminant` stablisation.

**Comment by jswrenn** [comment]
We're going to need to state *somewhere* that future versions of Rust *might* allow `as`-casting fieldless-but-non-c-like enums with explicit reprs. Rust programmers cannot count on these enums *always* being non-as-castable (and, consequently, that their reprs will remain unobservable in safe code).

If we don't make this clear, then we cannot revert #89234 in favor of #88203 in the future without introducing breaking backcompat. (Because changes that aren't semver-breaking in one rustc version would become semver-breaking in future rustc versions.)

**Comment by CeleritasCelery** [comment]
How would making fieldless-but-non-c-like enums `as`-castable be a breaking change?

**Comment by jswrenn** [comment]
> How would making fieldless-but-non-c-like enums `as`-castable be a breaking change?

If a library author defines a type like:
```rust
enum AnswerTo {
    TheQuestion{} = 42,
}
```
it's reasonable for a dependents of that library to assume that `assert_eq!(AnswerTo::TheQuestion as u8, 42)` will hold across non-major versions of that library. The library author should consider changing the discriminant of that variant to be a breaking change.

However, that same library might think they don't have to worry about changing the discriminant of *this* enum:
```rust
#[repr(u8)]
enum AnswerTo {
    TheQuestion{} = 42,
}
```
...because #89234 makes it **not** as-castable. The library author can change the discriminant to whatever they'd like, and can safely assume they won't break downstream users.

But, if we revert #89234 in favor of #88203, this is no longer a safe assumption to make. Downstream users *could* start as-casting `AnswerTo`, depending on what rustc version they are using.

**Comment by Soveu** [comment]
One note:
```rust
#[repr(u8)]
enum Enum {
  A = 1,
  B(u32) = 2,
}
```
In this case, `Enum::B` could be treated in two ways - as `2u8` or as a `fn(u32) -> Enum`

**Comment by joshtriplett** [comment]
There are a set of changes related to enums that we need to make, including discriminants, `as` casts, `AsRepr`, and similar. Lang has had a design meeting on that, and needs to have another once someone supplies a complete design that fits these pieces together.

**Comment by jswrenn** [comment]
Would anyone like to collaborate with me on this design? I'm very familiar with these issues, and have some design sketches on hand, but would love a collaborator help construct a cohesive vision of the future.

**Comment by ehuss** [comment]
I'm going to close as I believe this is now essentially complete as of 1.66.

There might be some longer-term decisions to make around `as` casting.  There are some strange exceptions:

* Fieldless enums with variants such as `Tuple()` or `Struct{}` allow `as` casting without a `#[repr(…)]`, and have since 1.15.

```rust
// Allowed since 1.15, even without a #[repr]
enum Fieldless {
    Tuple(),
    Struct{},
    Unit,
}
assert_eq!(0, Fieldless::Tuple() as isize);
assert_eq!(1, Fieldless::Struct{} as isize);
assert_eq!(2, Fieldless::Unit as isize);
```

* Similar to above, fieldless enums allow `as` casting as long as you don't specify an explicit discriminant.  Using an explicit discriminant suddenly prevents all `as` casts.

* Fieldless enums allow explicit discriminants on unit variants, but not struct or tuple (new 1.66). It's not clear to me why it was decided to allow this.

```rust
#[repr(u8)]
enum FieldlessWithDiscrimants {
    First = 10,
    Tuple(),
    Second = 20,
    Struct{},
    Unit,
}
assert_eq!(10, FieldlessWithDiscrimants::First as u8);
assert_eq!(11, FieldlessWithDiscrimants::Tuple() as u8);
assert_eq!(20, FieldlessWithDiscrimants::Second as u8);
assert_eq!(21, FieldlessWithDiscrimants::Struct{} as u8);
assert_eq!(22, FieldlessWithDiscrimants::Unit as u8);
```

I suspect further changes may require an RFC or some design decisions with the lang team.

**Comment by elichai** [comment]
Is there a way to get the discriminant of these enums without constructing one? something like the following:
```rust
#[repr(u8)]
enum MyValue {
    A(usize) = 0,
    B(u8) = 1,
}
let a = MyValue::A(_) as u8;
```
